### PR TITLE
Compilation of when expressions with redundant branches and guards

### PR DESCRIPTION
**Pull request by ayazhafiz**
- [Ensure that when jumping to a branch, all pattern symbols are loaded](https://github.com/roc-lang/roc/commit/34045d78116221f156a72d55c1b6a81303ec08e7)

If we are jumping to a target branch, it is necessary that the target
branch has all required pattern symbols loaded in it. Usually this is
already the case, but there is an exception with guarded patterns.
Guarded patterns have their patterns loaded only right before the guard
is evaluated, which happens at some point further along the decision
tree. As such, when a guarded pattern jumps to its target destination,
it should append the loaded patterns as parameters on the target
joinpoint.

- [Handle guards that appear multiple times in a compiled decision tree](https://github.com/roc-lang/roc/commit/d5a5ae97515cff4e1ce0a38e33bc5487cc128c39)

Suppose we have a when expression

```
15 if foo -> <b1>
b  if bar -> <b2>
_         -> <b3>
```

that may have a decision tree like

```
15?
  \true => foo?
              \true  => <b1>
              \false => bar?
                           \true  => <b2>
                           \false => <b3>
  \false => bar?
               \true  => <b2>
               \false => <b3>
```

In this case, the guard "bar?" appears twice in the compiled decision
tree. We need to materialize the guard expression in both locations in
the compiled tree, which means we cannot as-is stamp a compiled `bar?`
twice in each location. The reason is that

- the compiled joinpoint for each `bar?` guard needs to have a unique ID
- the guard expression might have call which needs unique call spec IDs,
  or other joins that need unique joinpoint IDs.

So, save the expression as we build up the decision tree and materialize
the guard each time we need it. In practice the guards should be quite
small, so duplicating should be fine. We could avoid duplication, but
it's not clear to me how to do that exactly since the branches after the
guard might end up being different.
@[ayazhafiz](https://github.com/roc-lang/roc/commits?author=ayazhafiz)
ayazhafiz committed 12 minutes ago
Verified

Closes https://github.com/roc-lang/roc/issues/5176
Closes https://github.com/roc-lang/roc/issues/5193

### CODE REVIEW

File: crates/compiler/mono/src/ir.rs [hunk 1]
```diff
@@ -9486,6 +9567,62 @@ impl<'a> Pattern<'a> {
 
         false
     }
+
+    // TODO: vast majority of the time, the patterns will be singleton or empty.
+    // We should introduce a smallvec optimized for the singleton case.
```
Comment (folkertdev) [inline crates/compiler/mono/src/ir.rs:9572]: can we make this return an `impl Iterator` maybe?

**Review by folkertdev — COMMENTED** [review]
Looks good, but it touches some mono output, so I think we should rebase this on current main

File: crates/compiler/mono/src/ir.rs (hunk 1 above)
Comment (ayazhafiz) [reply crates/compiler/mono/src/ir.rs:9572]: yeah, I didn't want to implement the iterator - but we can go ahead and do that.

File: crates/compiler/mono/src/ir/pattern.rs [hunk 2]
```diff
@@ -0,0 +1,1815 @@
...
+                        PatternBindingWork::RecordDestruct(_) => todo!(),
+                    }
+                }
+
+                *self = Done;
+                None
+            }
+        }
+    }
+}
```
Comment (ayazhafiz) [inline crates/compiler/mono/src/ir/pattern.rs:261]: @folkertdev this is the new iterator. I went ahead and moved a bunch of the pattern stuff to a new file.

**Review by folkertdev — APPROVED** [review]
nice!

**Pull request merged**